### PR TITLE
Implement Prio3MultiHotHistogram that checks bounded number of 1s

### PR DESCRIPTION
**Pull request by junyechen1996**
Add a reference implementation of the idea in issue #287, which needs a new FLP 'MultiHotHistogram' and a new Prio3 type 'Prio3MultiHotHistogram' that checks a Client's measurement has a bounded number of 1s.

### CODE REVIEW

**Review by wangshan — APPROVED** [review]

File: poc/vdaf_prio3.py [hunk 1]
```diff
@@ -483,10 +483,34 @@ class Prio3HistogramWithLength(Prio3Histogram):
         return Prio3HistogramWithLength
 
 
+class Prio3MultiHotHistogram(Prio3):
+    # Generic types required by `Prio3`
+    Xof = xof.XofShake128
+
+    # Associated parameters.
+    VERIFY_KEY_SIZE = xof.XofShake128.SEED_SIZE
+    ID = 0x00000FFF  # private codepoint for Prio3 types
```
Comment (cjpatton) [inline poc/vdaf_prio3.py:492]: This codepoint space is intended for Prio3 codepoints that are standardized. Let's leave it out until we've decided to do so. (If it turns out that we do want this for the DP draft, and the DP draft gets adopted, then we will certainly add it to the spec!)
```suggestion
    ID = 0xFFFFFFFF  # private codepoint for Prio3 types
```

File: poc/flp_generic.py [hunk 2]
```diff
@@ -640,6 +641,140 @@ def test_vec_set_type_param(self, test_vec):
         return ['length', 'chunk_length']
 
 
+class MultiHotHistogram(Valid):
+    """
+    A validity circuit that checks each Client's measurement is a bit vector,
+    and has at most `max_num_bits` 1s.
```
Comment (cjpatton) [inline poc/flp_generic.py:647]:
```suggestion
    A validity circuit that checks each Client's measurement is a bit vector
    with at most `max_num_bits` 1s.
```

File: poc/flp_generic.py [hunk 3]
```diff
@@ -640,6 +641,140 @@ def test_vec_set_type_param(self, test_vec):
...
+
+    In order to check the Client measurement `x` has at most `max_num_bits` 1s,
```
Comment (cjpatton) [inline poc/flp_generic.py:649]: To match `valid()`
```suggestion
    In order to check the Client measurement `meas` has at most `max_num_bits` 1s,
```

File: poc/flp_generic.py [hunk 4]
```diff
@@ -640,6 +641,140 @@ def test_vec_set_type_param(self, test_vec):
...
+    we ask the Client to send an additional `num_bits_for_count` bits that
```
Comment (cjpatton) [inline poc/flp_generic.py:650]: Consistency with `SumVec` and `Sum`
```suggestion
    we ask the Client to send an additional `bits_for_count` bits that
```

File: poc/flp_generic.py (hunk 3 above)
Comment (cjpatton) [inline poc/flp_generic.py:649]: Self-consistency: We call this the "count check" in the circuit.
```suggestion
    In order to check the Client measurement `x` has at most `max_count` 1s,
```

File: poc/flp_generic.py [hunk 5]
```diff
@@ -640,6 +641,140 @@ def test_vec_set_type_param(self, test_vec):
...
+      bits.
+    - We can naturally bound `count` as the following:
+      `0 <= count <= 2**num_bits_for_count - 1`, and therefore:
+      `-offset <= \sum_i x_i <= max_num_bits`.
+    - Since we also verify each `x_i` is a bit, we can lower bound the summation
+      by 0. Therefore, we will be able to verify
+      `0 <= \sum_i x_i <= max_num_bits`.
+    """
+    # Operational parameters
+    length = None  # Set by `MultiHotHistogram.with_params()`
```
Comment (cjpatton) [inline poc/flp_generic.py:664]: Here and below.
```suggestion
    length = None  # Set by the constructor
```

File: poc/flp_generic.py [hunk 6]
```diff
@@ -640,6 +641,140 @@ def test_vec_set_type_param(self, test_vec):
...
+    max_num_bits = None  # Set by `MultiHotHistogram.with_params()`
+    chunk_length = None  # Set by `MultiHotHistogram.with_params()`
+
+    # Associated types
+    Measurement = Unsigned
```
Comment (cjpatton) [inline poc/flp_generic.py:669]: What is this type? For `Histogram` it's a single bucket index, but for `MultiHotHistogram` we need to be able to represent multiple locations being set.

File: poc/flp_generic.py [hunk 7]
```diff
@@ -640,6 +641,140 @@ def test_vec_set_type_param(self, test_vec):
...
+            (length + self.num_bits_for_count + chunk_length - 1) //
+            chunk_length
+        ]
+        self.MEAS_LEN = self.length + self.num_bits_for_count
+        self.OUTPUT_LEN = self.length
+
+    def eval(self, meas, joint_rand, num_shares):
+        self.check_valid_eval(meas, joint_rand)
+
+        # Check that each bucket is one or zero.
```
Comment (cjpatton) [inline poc/flp_generic.py:720]: Is this sufficient for the DP use case? Suppose our algorithm for adding noise is something like this:
```python3
for bucket_index in len(meas):
   meas[bucket_index] += sample_independent_noise()
```
Depending on what `sample_independent_noise()` does, `meas[bucket_index]` might be something other than `0` or `1`.

File: poc/flp_generic.py [hunk 8]
```diff
@@ -640,6 +641,140 @@ def test_vec_set_type_param(self, test_vec):
...
+                    meas_elem = self.Field(0)
+
+                inputs[j * 2] = r_power * meas_elem
+                inputs[j * 2 + 1] = meas_elem - shares_inv
+
+                r_power *= r
+
+            range_check += self.GADGETS[0].eval(self.Field, inputs)
+
+        # Check that `offset + \sum_{i=0}^{length-1} meas[i]` is equal to the
```
Comment (cjpatton) [inline poc/flp_generic.py:741]: Prefer simpler language in comments
```suggestion
        # Check that `offset` plus the some of the buckets is equal to the
```

**Review by cjpatton — COMMENTED** [review]
Great work! One clarifying question about the requirements, see inline comment.

File: poc/flp_generic.py (hunk 3 above)
Comment (junyechen1996) [reply poc/flp_generic.py]: Fixed at all sites for `x`.

File: poc/flp_generic.py (hunk 4 above)
Comment (junyechen1996) [reply poc/flp_generic.py]: Changed it at all locations.

**Review by junyechen1996 — COMMENTED** [review]
@cjpatton Please also feel free to add David to review, since it's his idea. I don't have permission to add him.

File: poc/flp_generic.py (hunk 6 above)
Comment (cjpatton) [reply poc/flp_generic.py]: Perhaps it should be a set of bucket indices, i.e., `set[Unsigned]`? That would line up better with the measurement type for `Histogram`.

File: poc/flp_generic.py (hunk 7 above)
Comment (cjpatton) [reply poc/flp_generic.py:717]: Great, thanks for clarifying.

File: poc/flp_generic.py [hunk 9]
```diff
@@ -640,6 +641,137 @@ def test_vec_set_type_param(self, test_vec):
...
+      summation by 0. Therefore, we will be able to verify
+      `0 <= \sum_i meas_i <= max_count`.
+    """
+    # Operational parameters
+    length = None  # Set by the constructor
+    max_count = None  # Set by constructor
+    chunk_length = None  # Set by constructor
+
+    # Associated types
+    Measurement = Vec[Unsigned]
```
Comment (cjpatton) [inline poc/flp_generic.py:669]: Idea: the measurement is the set of bucket indices.
```suggestion
    Measurement = set[Unsigned]
```

File: poc/flp_generic.py [hunk 10]
```diff
@@ -640,6 +641,137 @@ def test_vec_set_type_param(self, test_vec):
...
+        # Check that `offset` plus the sum of the buckets is equal to the
+        # value claimed by the Client.
+        count_check = self.offset * shares_inv
+        for i in range(self.length):
+            count_check += meas[i]
+        count_check -= self.Field.decode_from_bit_vector(meas[self.length:])
+
+        out = joint_rand[1] * range_check + \
+            joint_rand[1] ** 2 * count_check
+        return out
```
Comment (cjpatton) [inline poc/flp_generic.py:747]:
```suggestion
        return joint_rand[1]    * range_check + \
               joint_rand[1]**2 * count_check
```

**Review by cjpatton — COMMENTED** [review]
One minor suggestion for the measurement type.

File: poc/flp_generic.py [hunk 11]
```diff
@@ -640,6 +641,137 @@ def test_vec_set_type_param(self, test_vec):
...
+        """
+        if length <= 0:
+            raise ValueError('invalid length')
+        if max_count <= 0 or max_count > length:
+            raise ValueError('invalid max_count')
+        if chunk_length <= 0:
+            raise ValueError('invalid chunk_length')
+
+        # Compute the number of bits to represent `max_count`.
+        self.bits_for_count = int(math.log2(max_count)) + 1
```
Comment (divergentdave) [inline poc/flp_generic.py:693]: We may as well avoid floating point math when computing this, so we don't have to worry about precision and rounding.
```suggestion
        self.bits_for_count = max_count.bit_length()
```

File: poc/flp_generic.py (hunk 10 above)
Comment (divergentdave) [reply poc/flp_generic.py:747]: We have `autopep8` in our linter CI job, and it won't like the irregular number of spaces around operators. We could add some extra parentheses if readability is of concern.

**Review by divergentdave — APPROVED** [review]

File: poc/flp_generic.py (hunk 10 above)
Comment (cjpatton) [reply poc/flp_generic.py:747]: :(((

No no, let's do what the robot tells us to do. The main thing is to be consistent I think.

**Review by junyechen1996 — COMMENTED** [review]
Rebased on main. Also squashed commits because I have to resolve conflicts.

File: poc/flp_generic.py (hunk 9 above)
Comment (cjpatton) [reply poc/flp_generic.py]: I don't have a strong preference.
* Even if you go with `set`, I don't think there's a need to change the API in the DP draft to match: just map one type to the other.
* If you stick with what you have please (1) change `Vec` to `list` and (2) add a comment describing the measurement type.

Comment (junyechen1996) [reply poc/flp_generic.py]: Sounds good. I changed it to `list[Unsigned]` and added comments next to `Measurement` type and `AggResult` type.

**Review by cjpatton — APPROVED** [review]

**Comment by junyechen1996** [comment]
Squashed.

**Pull request merged**